### PR TITLE
Distributed data parallelism

### LINKED ISSUE #2: Run with distributed parallelism
Add `torch.nn.parallel.DistributedDataParallel`. Current plan:

- [x] Replace `DataParallel` calls with `DDP`
~~Inherit from `Dataset` for data~~
- [ ] Wrap data with `DistributedSampler`
- [ ] Add 2 rank test

### PR DESCRIPTION
**Pull request by allaffa**
@pzhanggit @streeve 
I assign you as reviewers for my pull request that runs DistributedDataParallelismn on GCNN
I tested the code also on my DGX machine to make sure that the distributed processes take advantage of the GPUs when needed. 

Closes #2

### CODE REVIEW

File: utils/utils.py [hunk 1]
```diff
@@ -65,14 +86,24 @@ def train_validate_test_normal(
 
 
 def train(loader, model, opt, output_dim):
-    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
-    total_error = 0
+
+    if isinstance(model, torch.nn.parallel.distributed.DistributedDataParallel):
+        device = next(model.module.parameters()).device
```
Comment (streeve) [inline utils/utils.py:91]: This is an annoying API difference

File: utils/utils.py [hunk 2]
```diff
@@ -65,14 +86,24 @@ def train_validate_test_normal(
...
+
     model.train()
+
+    total_error = 0
     for data in tqdm(loader):
         data = data.to(device)
         opt.zero_grad()
-        pred = model(data)
-        loss = model.loss_rmse(pred, data.y)
+        if isinstance(model, torch.nn.parallel.distributed.DistributedDataParallel):
```
Comment (streeve) [inline utils/utils.py:101]: Importing directly could shorten this quite a bit

File: utils/utils.py [hunk 3]
```diff
@@ -173,42 +232,108 @@ def dataset_loading_and_splitting(
...
+
+    trainset = dataset[: int(data_size * perc_train)]
+    valset = dataset[
+        int(data_size * perc_train) : int(data_size * (perc_train + perc_val))
+    ]
+    testset = dataset[int(data_size * (perc_train + perc_val)) :]
+
+    if distributed_data_parallelism:
+
+        train_sampler = torch.utils.data.distributed.DistributedSampler(trainset)
```
Comment (streeve) [inline utils/utils.py:256]: So you can still wrap the existing data with `DistributedSampler` without inheriting?

File: utils/utils.py [hunk 4]
```diff
@@ -173,42 +232,108 @@ def dataset_loading_and_splitting(
...
+            batch_size=batch_size,
+            shuffle=True,
+        )
 
     return train_loader, val_loader, test_loader
 
 
 def combine_and_split_datasets(
-    dataset1: [], dataset2: [], batch_size: int, perc_train: float
+    dataset1: [],
```
Comment (streeve) [inline utils/utils.py:288]: May as well rename these `trainset` and `testset` (unless you're worried about backwards compatibility)

File: utils/utils.py [hunk 5]
```diff
@@ -173,42 +232,108 @@ def dataset_loading_and_splitting(
...
-        shuffle=True,
-    )
+
+    trainset = dataset[: int(data_size * perc_train)]
+    valset = dataset[
+        int(data_size * perc_train) : int(data_size * (perc_train + perc_val))
+    ]
+    testset = dataset[int(data_size * (perc_train + perc_val)) :]
+
+    if distributed_data_parallelism:
```
Comment (streeve) [inline utils/utils.py:254]: This section for splitting into test/train/validation for parallel/serial should be a separate function called by both `split_dataset` and `combine_and_split_datasets`

File: utils/models_setup.py [hunk 6]
```diff
@@ -8,9 +9,47 @@
...
+def generate_model(
+    model_type: str,
+    input_dim: int,
+    dataset: [Data],
+    config: dict,
+    distributed_data_parallelism: bool = False,
+):
+
+    if distributed_data_parallelism:
+        world_size = os.environ["OMPI_COMM_WORLD_SIZE"]
```
Comment (streeve) [inline utils/models_setup.py:43]: I would extract this the same way as initialization, so maybe a separate function called within `setup`?

File: run_config_input.py [hunk 7]
```diff
@@ -216,6 +220,17 @@ def run_normal_config_file():
         + str(config["predicted_value_option"])
     )
 
+    if run_in_parallel:
+        available_gpus = get_gpus_list()
+        model = torch.nn.parallel.DistributedDataParallel(
+            model, device_ids=["cuda:" + str(int(world_rank) % len(available_gpus))]
```
Comment (streeve) [inline run_config_input.py:226]: This breaks on CPU. And in general, shouldn't we call `get_gpu` (`get_device`) here instead of doing this directly?

File: utils/models_setup.py [hunk 8]
```diff
@@ -8,9 +9,47 @@
...
+        world_rank = os.environ["OMPI_COMM_WORLD_RANK"]
+
     torch.manual_seed(0)
-    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+    available_gpus = get_gpus_list()
+    if len(available_gpus) > 0:
+        device = get_gpu(int(world_rank) % len(available_gpus))
```
Comment (streeve) [inline utils/models_setup.py:50]: I think that calculation should be done inside `get_gpu`

**Review by streeve — COMMENTED** [review]
Just a first round of review to try to understand and clean up the interfaces. 

Having a simple class for parallel details that we pass around could simplify a lot (it could hold a `bool` for whether we're running in parallel, `world_size`, etc.)

**Comment by streeve** [comment]
Also, per #2 we'll want a test that runs in parallel before we merge

File: distributed_data_parallelism_main.py [hunk 9]
```diff
@@ -0,0 +1,18 @@
+import os
+import argparse
+
+
+from run_config_input import run_normal_config_file
+
+
+def main():
```
[the remark this answers is not in the document]
Comment (allaffa) [reply distributed_data_parallelism_main.py:8]: Yes, but you cannot allow the user to use an interactive input selection if you want to to run things in parallel

File: run_config_input.py [hunk 10]
```diff
@@ -156,6 +157,12 @@ def run_normal_terminal_input():
 
 
 def run_normal_config_file():
+
+    run_in_parallel = setup_ddp()
+    if run_in_parallel:
+        world_size = os.environ["OMPI_COMM_WORLD_SIZE"]
```
[the remark this answers is not in the document]
Comment (allaffa) [reply run_config_input.py]: @streeve I push ed changes to take care of this.

File: utils/models_setup.py [hunk 11]
```diff
@@ -8,9 +9,47 @@
 from models.MFCStack import MFCStack
 
 
-def generate_model(model_type: str, input_dim: int, dataset: [Data], config: dict):
+def get_gpus_list():
+
+    available_gpus = []
```
[the remark this answers is not in the document]
Comment (allaffa) [reply utils/models_setup.py]: removed

File: utils/models_setup.py [hunk 12]
```diff
@@ -8,9 +9,47 @@
...
+def get_gpu(number):
+    gpus_list = get_gpus_list()
+    if number not in gpus_list:
+        raise ValueError(
+            "The GPU ID:" + str(number) + " is not inside the list of GPUs available"
+        )
+    else:
+        device = torch.device(
+            "cuda:" + str(number)
+        )  # you can continue going on here, like cuda:1 cuda:2....etc.
```
[the remark this answers is not in the document]
Comment (allaffa) [reply utils/models_setup.py:28]: This is needed to grab the "device" with a specific index. The function `torch.device()` takes as an input argument 
a string like 'cuda:number'

File: utils/models_setup.py (hunk 8 above)
Comment (allaffa) [reply utils/models_setup.py:49]: get_gpu returns the torch.device object associated with a specific number ID, assuming that the number ID is in the liste of the available CUDA devices, otherwise it returns an error. 
In view of a potential extension of the code where a single neural network does not fit on a single GPU, what you propose may be limiting.

File: utils/utils.py (hunk 4 above)
Comment (allaffa) [reply utils/utils.py:290]: actually there are 3 partitioning performed on 2 datasets

dataset1 is split into training and validation. dataset 2 becomes the testset. 
So renaming dataset1 as trainingset may be confusing

File: run_config_input.py (hunk 7 above)
Comment (allaffa) [reply run_config_input.py]: Changed as suggested

File: utils/utils.py (hunk 5 above)
Comment (allaffa) [reply utils/utils.py:242]: changed as suggested

**Comment by streeve** [comment]
@allaffa I updated to run on cpu as well and added an mpi test. This should pass when we rebase on #18

**Comment by allaffa** [comment]
@streeve Is it normal that the tests do not pass? i think you did not run black before pushing the last changes.

**Comment by streeve** [comment]
One question is a good way to test correctness for these runs - right now it just makes sure the code runs without errors, but maybe we leave that to the next PR

**Review by streeve — APPROVED** [review]

**Comment by streeve** [comment]
Also, @pzhanggit there are "approve"/"request changes" options in the code changes tab

**Comment by streeve** [comment]
@allaffa merge when you like

**Comment by allaffa** [comment]
@streeve done
